crates/node/service/src/actors/sequencer/origin_selector.rs

```diff
@@ -177,20 +178,57 @@ pub trait L1OriginSelectorProvider {
     ) -> Result<Option<BlockInfo>, L1OriginSelectorError>;
 }
 
+/// A wrapper around the [`RootProvider`] that delays the view of the L1 chain by a configurable
+/// amount of blocks.
+#[derive(Debug)]
+pub struct DelayedL1OriginSelectorProvider {
+    /// The inner [`RootProvider`].
+    inner: RootProvider,
+    /// The L1 head watch channel.
+    l1_head: watch::Receiver<Option<BlockInfo>>,
+    /// The confirmation depth to delay the view of the L1 chain.
+    confirmation_depth: u64,
+}
+
+impl DelayedL1OriginSelectorProvider {
+    /// Creates a new [`DelayedL1OriginSelectorProvider`].
+    pub const fn new(
+        inner: RootProvider,
+        l1_head: watch::Receiver<Option<BlockInfo>>,
+        confirmation_depth: u64,
+    ) -> Self {
+        Self { inner, l1_head, confirmation_depth }
+    }
+}
+
 #[async_trait]
-impl L1OriginSelectorProvider for RootProvider {
+impl L1OriginSelectorProvider for DelayedL1OriginSelectorProvider {
     async fn get_block_by_hash(
         &self,
         hash: B256,
     ) -> Result<Option<BlockInfo>, L1OriginSelectorError> {
-        Ok(Provider::get_block_by_hash(self, hash).await?.map(Into::into))
+        // By-hash lookups are not delayed, as they're direct indexes.
+        Ok(Provider::get_block_by_hash(&self.inner, hash).await?.map(Into::into))
     }
 
     async fn get_block_by_number(
         &self,
         number: u64,
     ) -> Result<Option<BlockInfo>, L1OriginSelectorError> {
-        Ok(Provider::get_block_by_number(self, number.into()).await?.map(Into::into))
+        let l1_head = *self.l1_head.borrow();
+        match l1_head {
+            Some(l1_head)
+                if number == 0 ||
+                    self.confirmation_depth == 0 ||
+                    number + self.confirmation_depth <= l1_head.number =>
+            {
+                Ok(Provider::get_block_by_number(&self.inner, number.into()).await?.map(Into::into))
+            }
+            None => {
+                Ok(Provider::get_block_by_number(&self.inner, number.into()).await?.map(Into::into))
+            }
```

Comment:
Personally would prefer if this was in a `let Some(l1_head) = *self.l1_head.borrow() else { ... };` pattern to simplify the branching but not a big deal.